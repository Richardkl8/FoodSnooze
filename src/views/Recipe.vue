<template>
    <div class="container">
        <section id="instructions">
            <div class="row my-5">
                <div class="col-sm-12 col-md-6">
                    <img :src="recipe.strMealThumb"
                         alt="recipe-image"
                         class="recipe-image">
                </div>
                <div class="col-sm-12 col-md-6">
                    <custom-h2>
                        <b>
                            {{recipe.strMeal}}
                        </b>
                    </custom-h2>
                    <custom-h3>
                        {{recipe.strInstructions}}
                    </custom-h3>
                </div>
            </div>
        </section>
        <section id="ingredients">
            <div class="ingredients my-5">
                <custom-h2 class="mb-4"><b>Ingredients</b></custom-h2>
                <div class="row">
                    <div v-for="ingredient in returnListOfIngredients()"
                         :key="ingredient"
                         class="col-md-6">
                        <custom-h3 class="text-black">
                            <ul>
                                <li>
                                    {{ingredient}}
                                    <hr>
                                </li>
                            </ul>
                        </custom-h3>
                    </div>
                </div>
            </div>
        </section>
        <section v-if="videoUrl" id="video">
            <custom-h2 class="mb-4"><b>Checkout the video</b></custom-h2>
            <div class="row my-5">
                <div class="col">
<!--                    TODO: create class for this-->
                    <iframe id="recipe-video"
                            :src="videoUrl"
                            frameborder="0" allowfullscreen>
                    </iframe>
                </div>
            </div>
        </section>
        <section id="relatedFoods">
            <custom-h2 class="mb-4"><b>You may also like</b></custom-h2>
            <div class="row my-5">
                <div v-for="index in 3"
                     :key="index"
                     class="col-xs-12 col-md-4">
                    <recipe-card
                        :recipe="relatedMeals[index]"
                        @click.native="$_goToRecipePage(relatedMeals[index].idMeal)">
                    </recipe-card>
                </div>
            </div>
        </section>
    </div>
</template>

<script>
    import RecipeCard from '@/components/RecipeCard.vue';
    import goToPageMixin from '../mixins/goToPageMixin';

    export default {
        name: 'Home',
        components: {
            RecipeCard,
        },
        mixins: [
            goToPageMixin,
        ],
        inject: [
            'apiService',
        ],
        data() {
            return {
                recipe: {},
                relatedMeals: [],
                videoUrl: '',
            };
        },
        watch: {
            $route() {
                this.getRecipeById();
            },
        },
        methods: {
            getRecipeById() {
                this.apiService.getRecipeById(this.$route.params.id)
                    .then((response) => {
                        this.recipe = response;
                        this.getRecipeByCategory(response.strCategory);
                        this.setYoutubeEmbedUrl(response.strYoutube);
                    });
            },
            getRecipeByCategory(category) {
                this.apiService.getRecipeByCategory(category)
                    .then((response) => {
                        this.relatedMeals = response;
                    });
            },
            // Youtube does not let you play video on your website without embedding them,
            // Therefore the url needs to be edited
            setYoutubeEmbedUrl() {
                this.videoUrl = this.recipe.strYoutube.replace('/watch?v=', '/embed/');
            },
            returnListOfIngredients() {
                const entries = Object.entries(this.recipe);
                const ingredients = [];
                const measures = [];
                const final = [];

                entries.forEach(([key, value]) => {
                    if (key.startsWith('strIngredient') && value) {
                        ingredients.push(value);
                    }
                    if (key.startsWith('strMeasure') && value) {
                        measures.push(value);
                    }
                });

                ingredients.forEach((ingredient, index) => {
                    if (measures[index] !== ' ') {
                        final.push(`${ingredient} - ${measures[index]}`);
                    } else {
                        final.push(ingredient);
                    }
                });
                return final;
            },
        },
        created() {
            this.getRecipeById();
        },
    };
</script>

<style lang="scss" scoped>
    .recipe-image {
        max-width: 100%;
    }

    ul {
        padding: 0;
        margin: 0;
        list-style-type: none;
    }

    li {
        margin: 1rem 0;
    }

    #recipe-video {
        width: 100%;
        height: 60rem;
    }

    .ingredients {
        padding: 3rem 2rem;
        border: solid 2px #2DB396;
        color: #fff;
    }

    @media screen and (max-width: 768px) {
        #recipe-video {
            height: 30rem;
        }
    }

    @media screen and (max-width: 1200px) {
        #recipe-video {
            height: 45rem;
        }
    }
</style>
