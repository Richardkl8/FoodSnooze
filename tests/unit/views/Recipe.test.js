import { shallowMount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Recipe from '../../../src/views/Recipe.vue';
import stubs from '../stubs/stubs';

const responseGetRecipeById = {
    idMeal: '12',
    strMealThumb: 'image.jpg',
    strMeal: 'tuna',
    strCategory: 'fish',
    strIngredient1: 'tuna',
    strIngredient2: 'lemon',
    strIngredient3: 'salt',
    strMeasure1: '1',
    strMeasure2: '2',
    strMeasure3: ' ',
    strInstructions: 'Bake the tuna',
    strYoutube: 'https://www.youtube.com/watch?v=1ahpSTf_Pvk',
};

const responseGetRecipeByCategory = [
    {
        strMeal: 'tuna',
        strMealThumb: 'image.jpg',
        idMeal: '123',
    },
    {
        strMeal: 'beef',
        strMealThumb: 'image2.jpg',
        idMeal: '321',
    },
];

const provide = {
    apiService: {
        getRecipeById: jest.fn().mockResolvedValue(responseGetRecipeById),
        getRecipeByCategory: jest.fn().mockResolvedValue(responseGetRecipeByCategory),
    },
};

const mocks = {
    $route: {
        params: {
            id: '123',
        },
    },
    $router: {
        push: jest.fn(),
    },
};


describe('Recipe page', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(Recipe, {
            provide,
            stubs,
            mocks,
        });
    });

    test('Should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Should call getRecipeById, getRecipeByCategorySpy and setYoutubeEmbedUrl when component is created', async () => {
        const getRecipeByIdSpy = jest.spyOn(Recipe.methods, 'getRecipeById');
        const getRecipeByCategorySpy = jest.spyOn(Recipe.methods, 'getRecipeByCategory');
        const setYoutubeEmbedUrlSpy = jest.spyOn(Recipe.methods, 'setYoutubeEmbedUrl');
        const newWrapper = shallowMount(Recipe, {
            provide,
            stubs,
            mocks,
        });
        await flushPromises();

        expect(getRecipeByIdSpy).toHaveBeenCalledTimes(1);
        expect(newWrapper.vm.apiService.getRecipeById).toHaveBeenCalledWith('123');
        expect(getRecipeByCategorySpy).toHaveBeenCalledWith('fish');
        expect(setYoutubeEmbedUrlSpy).toHaveBeenCalledWith('https://www.youtube.com/watch?v=1ahpSTf_Pvk');
    });

    test('Should edit youtube string to embedded url when created', () => {
        expect(wrapper.vm.$data.videoUrl).toBe('https://www.youtube.com/embed/1ahpSTf_Pvk');
    });

    test('Should return a list of HTML stings with ingredients and measurements combined', () => {
        const result = [
            { ingredient: 'tuna', measure: '1' },
            { ingredient: 'lemon', measure: '2' },
            { ingredient: 'salt', measure: ' ' },
        ];
        expect(wrapper.vm.getIngredients()).toEqual(result);
    });

    test('Should not show video section when videoUrl is empty', async () => {
        expect(wrapper.find('#video').isVisible()).toBe(true);
        wrapper.vm.$data.videoUrl = '';
        await flushPromises();
        expect(wrapper.find('#video').exists()).toBe(false);
    });

    test('Should update route with mealId when input is different then current mealId in route', async () => {
        wrapper.vm.updateRouteWithMealId('123');
        expect(mocks.$router.push).not.toHaveBeenCalledWith('123');

        wrapper.vm.updateRouteWithMealId('321');
        expect(mocks.$router.push).toHaveBeenCalledWith('/recipe/321');
    });
});
