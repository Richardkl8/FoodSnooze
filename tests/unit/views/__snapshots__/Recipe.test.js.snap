// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should call getRecipeById when component is created 1`] = `
<div class="container">
  <section id="instructions">
    <div class="row my-5">
      <div class="col-sm-12 col-md-6"><img alt="recipe-image" class="recipe-image" src="image.jpg"></div>
      <div class="col-sm-12 col-md-6">
        <custom-h2-stub><b>
            tuna
          </b></custom-h2-stub>
        <custom-h2-stub>
          fish
        </custom-h2-stub>
      </div>
    </div>
  </section>
  <section id="ingredients">
    <div class="row my-5">
      <div class="col-12">
        <custom-h2-stub class="mb-4 mt-5"><b>Ingredients</b></custom-h2-stub>
        <div class="row">
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>tuna</b> <i>1</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>lemon</b> <i>2</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section id="how-to-cook">
    <div class="row">
      <div class="col-12">
        <custom-h2-stub class="mb-4"><b>How to cook</b></custom-h2-stub>
        <custom-h3-stub>
          Bake the tuna
        </custom-h3-stub>
      </div>
    </div>
  </section>
  <section id="video">
    <custom-h2-stub class="my-5"><b>Checkout the video</b></custom-h2-stub>
    <div class="row mb-5">
      <div class="col"><iframe id="recipe-video" src="https://www.youtube.com/embed/1ahpSTf_Pvk" frameborder="0" allowfullscreen="allowfullscreen"></iframe></div>
    </div>
  </section>
  <section id="relatedFoods">
    <custom-h2-stub class="mt-5 mb-4"><b>You may also like</b></custom-h2-stub>
    <div class="row my-5">
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub recipe="[object Object]"></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
    </div>
  </section>
</div>
`;

exports[`Home Should call getRecipeById, getRecipeByCategorySpy and setYoutubeEmbedUrl when component is created 1`] = `
<div class="container">
  <section id="instructions">
    <div class="row my-5">
      <div class="col-sm-12 col-md-6"><img alt="recipe-image" class="recipe-image" src="image.jpg"></div>
      <div class="col-sm-12 col-md-6">
        <custom-h2-stub><b>
            tuna
          </b></custom-h2-stub>
        <custom-h2-stub>
          fish
        </custom-h2-stub>
      </div>
    </div>
  </section>
  <section id="ingredients">
    <div class="row my-5">
      <div class="col-12">
        <custom-h2-stub class="mb-4 mt-5"><b>Ingredients</b></custom-h2-stub>
        <div class="row">
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>tuna</b> <i>1</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>lemon</b> <i>2</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section id="how-to-cook">
    <div class="row">
      <div class="col-12">
        <custom-h2-stub class="mb-4"><b>How to cook</b></custom-h2-stub>
        <custom-h3-stub>
          Bake the tuna
        </custom-h3-stub>
      </div>
    </div>
  </section>
  <section id="video">
    <custom-h2-stub class="my-5"><b>Checkout the video</b></custom-h2-stub>
    <div class="row mb-5">
      <div class="col"><iframe id="recipe-video" src="https://www.youtube.com/embed/1ahpSTf_Pvk" frameborder="0" allowfullscreen="allowfullscreen"></iframe></div>
    </div>
  </section>
  <section id="relatedFoods">
    <custom-h2-stub class="mt-5 mb-4"><b>You may also like</b></custom-h2-stub>
    <div class="row my-5">
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub recipe="[object Object]"></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
    </div>
  </section>
</div>
`;

exports[`Home Should match snapshot 1`] = `
<div class="container">
  <section id="instructions">
    <div class="row my-5">
      <div class="col-sm-12 col-md-6"><img alt="recipe-image" class="recipe-image" src="image.jpg"></div>
      <div class="col-sm-12 col-md-6">
        <custom-h2-stub><b>
            tuna
          </b></custom-h2-stub>
        <custom-h2-stub>
          fish
        </custom-h2-stub>
      </div>
    </div>
  </section>
  <section id="ingredients">
    <div class="row my-5">
      <div class="col-12">
        <custom-h2-stub class="mb-4 mt-5"><b>Ingredients</b></custom-h2-stub>
        <div class="row">
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>tuna</b> <i>1</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>lemon</b> <i>2</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section id="how-to-cook">
    <div class="row">
      <div class="col-12">
        <custom-h2-stub class="mb-4"><b>How to cook</b></custom-h2-stub>
        <custom-h3-stub>
          Bake the tuna
        </custom-h3-stub>
      </div>
    </div>
  </section>
  <section id="video">
    <custom-h2-stub class="my-5"><b>Checkout the video</b></custom-h2-stub>
    <div class="row mb-5">
      <div class="col"><iframe id="recipe-video" src="https://www.youtube.com/embed/1ahpSTf_Pvk" frameborder="0" allowfullscreen="allowfullscreen"></iframe></div>
    </div>
  </section>
  <section id="relatedFoods">
    <custom-h2-stub class="mt-5 mb-4"><b>You may also like</b></custom-h2-stub>
    <div class="row my-5">
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub recipe="[object Object]"></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
    </div>
  </section>
</div>
`;

exports[`Recipe page Should match snapshot 1`] = `
<div class="container">
  <section id="instructions">
    <div class="row my-5">
      <div class="col-sm-12 col-md-6"><img alt="recipe-image" class="recipe-image" src="image.jpg"></div>
      <div class="col-sm-12 col-md-6">
        <custom-h2-stub><b>tuna</b></custom-h2-stub>
        <custom-h2-stub>
          fish
        </custom-h2-stub>
      </div>
    </div>
  </section>
  <section id="ingredients">
    <div class="row my-5">
      <div class="col-12">
        <custom-h2-stub class="mb-4 mt-5"><b>Ingredients</b></custom-h2-stub>
        <div class="row">
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>tuna</b> <i>1</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
          <div class="col-md-6 col-sm-12">
            <custom-h3-stub class="text-black">
              <ul>
                <li><b>lemon</b> <i>2</i></li>
                <hr>
              </ul>
            </custom-h3-stub>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section id="how-to-cook">
    <div class="row">
      <div class="col-12">
        <custom-h2-stub class="mb-4"><b>How to cook</b></custom-h2-stub>
        <custom-h3-stub>
          Bake the tuna
        </custom-h3-stub>
      </div>
    </div>
  </section>
  <section id="video">
    <custom-h2-stub class="my-5"><b>Checkout the video</b></custom-h2-stub>
    <div class="row mb-5">
      <div class="col"><iframe id="recipe-video" src="https://www.youtube.com/embed/1ahpSTf_Pvk" frameborder="0" allowfullscreen="allowfullscreen"></iframe></div>
    </div>
  </section>
  <section id="relatedFoods">
    <custom-h2-stub class="mt-5 mb-4"><b>You may also like</b></custom-h2-stub>
    <div class="row my-5">
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub recipe="[object Object]"></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <recipe-card-stub></recipe-card-stub>
      </div>
    </div>
  </section>
</div>
`;
